File: python/resistor-color-duo/resistor_color_duo.py
# The band colors are encoded as follows:
# Black: 0
# Brown: 1
# Red: 2
# Orange: 3
# Yellow: 4
# Green: 5
# Blue: 6
# Violet: 7
# Grey: 8
# White: 9

RESISTANCE_VALUES = {
    "black": 0,
    "brown": 1,
    "red": 2,
    "orange": 3,
    "yellow": 4,
    "green": 5,
    "blue": 6,
    "violet": 7,
    "grey": 8,
    "white": 9,
}


def value(colors):
    colors = colors[0:2]  # Ignore additional colors
    return int("".join([str(RESISTANCE_VALUES[color]) for color in colors]))

File: go/resistor-color-duo/resistor_color_duo.go
package resistorcolorduo

var colorToResistance = map[string]int{
	"black":  0,
	"brown":  1,
	"red":    2,
	"orange": 3,
	"yellow": 4,
	"green":  5,
	"blue":   6,
	"violet": 7,
	"grey":   8,
	"white":  9,
}

// Value should return the resistance value of a resistor with a given colors.
func Value(colors []string) int {
	colors = trimExtraColors(colors)
	resistances := decodeColors(colors)
	return combine(resistances)
}

func trimExtraColors(colors []string) []string {
	// ignore extra colors because only two are supported
	return colors[:2]
}

func combine(resistances []int) int {
	if resistances[0] == 0 {
		return resistances[1]
	}
	return resistances[0]*10 + resistances[1]
}

func decodeColors(colors []string) (resistances []int) {
	for _, color := range colors {
		resistance := decode(color)
		resistances = append(resistances, resistance)
	}
	return resistances
}

func decode(color string) (resistance int) {
	return colorToResistance[color]
}
